lib/src/main/java/io/ably/lib/realtime/Channel.java

```diff
@@ -215,6 +228,155 @@ public void sync() throws AblyException {
+	synchronized private void clearAttachTimers() {
+		Timer[] timers = new Timer[]{attachTimer, reattachTimer};
+		attachTimer = reattachTimer = null;
+		for (Timer t: timers) {
+			if (t != null) {
+				t.cancel();
+				t.purge();
+			}
+		}
+	}
+
+	/**
+	 * Attach channel, if not attached within timeout set state to suspended and
+	 * set up timer to reattach it later
+	 */
+	synchronized private void attachWithTimeout(final CompletionListener listener) throws AblyException {
+		final Timer currentAttachTimer = new Timer();
+		attachTimer = currentAttachTimer;
+
+		try {
+			attachImpl(new CompletionListener() {
+				@Override
+				public void onSuccess() {
+					clearAttachTimers();
+					if (listener != null)
+						listener.onSuccess();
+				}
+
+				@Override
+				public void onError(ErrorInfo reason) {
+					clearAttachTimers();
+					if (listener != null)
+						listener.onError(reason);
+				}
+			});
+		} catch(AblyException e) {
+			attachTimer = null;
+		}
+
+		if(attachTimer == null)
+			/* operation has already succeeded or failed, no need to set the timer */
+			return;
+
+		attachTimer.schedule(
+				new TimerTask() {
+					@Override
+					public void run() {
+						String errorMessage = String.format("Attach timed out for channel %s", name);
+						Log.v(TAG, errorMessage);
+						synchronized (Channel.this) {
+							if(attachTimer != currentAttachTimer)
+								return;
+							attachTimer = null;
+							if(state == ChannelState.attaching) {
+								setSuspended(new ErrorInfo(errorMessage, 91200));
+								reattachAfterTimeout();
+							}
+						}
+					}
+				}, Defaults.realtimeRequestTimeout);
+	}
+
+	/**
+	 * Must be called in suspended state. Wait for timeout specified in clientOptions, and then
+	 * try to attach the channel
+	 */
+	synchronized private void reattachAfterTimeout() {
+		final Timer currentReattachTimer = new Timer();
+		reattachTimer = currentReattachTimer;
+
+		reattachTimer.schedule(new TimerTask() {
+			@Override
+			public void run() {
+				synchronized (Channel.this) {
+					if (currentReattachTimer != reattachTimer)
+						return;
+					reattachTimer = null;
+					if (state == ChannelState.detached) {
+						try {
+							attachWithTimeout(null);
+						} catch (AblyException e) {
+							Log.e(TAG, "Reattach channel failed; channel = " + name, e);
+						}
+					}
+				}
+			}
+		}, Defaults.TIMEOUT_CHANNEL_RETRY);
+	}
+
+	/**
+	 * Try to detach the channel. If the server doesn't confirm the detach operation within realtime
+	 * request timeout return channel to previous state
+	 */
+	synchronized private void detachWithTimeout(final CompletionListener listener) throws AblyException {
+		final ChannelState originalState = state;
+		final Timer currentDetachTimer = new Timer();
+		attachTimer = currentDetachTimer;
+
+		try {
+			detachImpl(new CompletionListener() {
+				@Override
+				public void onSuccess() {
+					clearAttachTimers();
+					if (listener != null)
+						listener.onSuccess();
```

Comment:
Completely aside from this, the listeners don't appear to conform to http://docs.ably.io/client-lib-development-guide/versions/v0.8/features/#RTE6 even though RTE6 only talks about events being emitted.  Should we add a new spec item that states all callbacks / listeners that raise an exception must be caught, logged, and execution should continue?